core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java

```diff
@@ -0,0 +1,220 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.jdbc;
+
+import java.sql.DataTruncation;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.SQLIntegrityConstraintViolationException;
+import java.sql.SQLNonTransientConnectionException;
+import java.sql.SQLTimeoutException;
+import java.sql.SQLTransientConnectionException;
+import java.sql.SQLWarning;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.iceberg.BaseMetastoreTableOperations;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class JdbcTableOperations extends BaseMetastoreTableOperations {
+
+  private static final Logger LOG = LoggerFactory.getLogger(JdbcTableOperations.class);
+  private final String catalogName;
+  private final TableIdentifier tableIdentifier;
+  private final FileIO fileIO;
+  private final JdbcClientPool connections;
+
+  protected JdbcTableOperations(JdbcClientPool dbConnPool, FileIO fileIO, String catalogName,
+                                TableIdentifier tableIdentifier) {
+    this.catalogName = catalogName;
+    this.tableIdentifier = tableIdentifier;
+    this.fileIO = fileIO;
+    this.connections = dbConnPool;
+  }
+
+  @Override
+  public void doRefresh() {
+    Map<String, String> table;
+
+    try {
+      table = getTable();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new UncheckedInterruptedException(e, "Interrupted during refresh");
+    } catch (SQLException e) {
+      // SQL exception happened when getting table from catalog
+      throw new UncheckedSQLException(e, "Failed to get table %s from catalog %s", tableIdentifier, catalogName);
+    }
+
+    // Table not exists AND currentMetadataLocation is not NULL!
+    if (table.isEmpty() && currentMetadataLocation() != null) {
+      throw new NoSuchTableException("Failed to get table %s from catalog %s" +
+          " maybe another process deleted it", tableIdentifier, catalogName);
+    }
+
+    // Table not exists in the catalog! metadataLocation is null here!
+    if (table.isEmpty()) {
+      refreshFromMetadataLocation(null);
+      return;
+    }
+
+    // Table exists but metadataLocation is null
+    if (table.getOrDefault(JdbcUtil.METADATA_LOCATION, null) == null) {
```

Comment:
This `get` is done twice. Can you refactor to use a variable? I think this should also be `IllegalStateException` rather than just a `RuntimeException` so you could do this:

```java
String newMetadataLocation = table.get(JdbcUtil.METADATA_LOCATION);
Preconditions.checkState(newMetadataLocation != null, "Invalid table %s: metadata location is null", tableIdentifier);

refreshFromMetadataLocation(newMetadataLocation);
```

Reply:
(Also, `getOrDefault` with a `null` default is equivalent to `get`.)

Reply:
updated